File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    vector<vector<int>> directions = {{1,0}, {0,1}, {-1,0}, {0,-1}};
    void bfs(int i, int j, vector<vector<char>> &grid, vector<vector<bool>> &vis) {
        queue<vector<int>> q;
        q.push({i,j});
        vis[i][j] = true;
        while(q.size() > 0) {
            vector<int> temp = q.front();
            q.pop();
            for(int i=0; i<directions.size(); i++) {
                int x = directions[i][0] + temp[0];
                int y = directions[i][1] + temp[1];
                if(x < grid.size() && x >= 0 && y >= 0 && y < grid[0].size() && !vis[x][y] && grid[x][y] == '1') {
                    vis[x][y] = true;
                    q.push({x, y});
                }
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        vector<vector<bool>> vis (grid.size(), vector<bool>(grid[0].size(), false));
        int ans = 0;
        for(int i=0; i<grid.size(); i++) {
            for(int j=0; j<grid[i].size(); j++) {
                if(!vis[i][j] && grid[i][j] == '1') {
                    ans++;
                    bfs(i, j, grid, vis);
                }
            }
        }
        return ans;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    int[][] directions = new int[][]{{1,0},{0,1},{-1,0},{0,-1}};
    void bfs(int i, int j, char[][] grid, boolean[][] vis) {
        Deque<int[]> queue = new ArrayDeque();
        vis[i][j] = true;
        queue.addLast(new int[]{i,j});
        while(!queue.isEmpty()) {
            int[] temp = queue.removeFirst();
            for(int[] dir:directions) {
                int x = temp[0] + dir[0];
                int y = temp[1] + dir[1];
                if(x >= 0 && x < grid.length && y >= 0 && y < grid[0].length && grid[x][y] == '1' && !vis[x][y]) {
                    vis[x][y] = true;
                    queue.addLast(new int[]{x,y});
                }
            }
        }
    }
    public int numIslands(char[][] grid) {
        boolean[][] vis = new boolean[grid.length][grid[0].length];
        int ans = 0;
        for(int i=0; i<grid.length; i++) {
            for(int j=0; j<grid[i].length; j++) {
                if(!vis[i][j] && grid[i][j] == '1') {
                    ans++;
                    bfs(i, j, grid, vis);
                }
                else {
                    vis[i][j] = true;
                }
            }
        }
        return ans;
    }
}